Feature/ops 1722 build and deploy antivirus

RalfNellesCap opened the pull request:
# Description

<!--
  This is a template to add as many information as possible to the pull request, to help reviewer and as a checklist for you. Points to remember are set in the comments, please read and keep them in mind:

    - Code should be self-explanatory and share your knowledge with others
    - Document code that is not self-explanatory
    - Think about bugs and keep security in mind
    - Write tests (Unit and Integration), also for error cases
    - Main logic should hidden behind the api, never trust the client
    - Visible changes should be discussed with the UX-Team from the begining of development; they also have to accept them at the end
    - Keep the changelog up-to-date
    - Leave the code cleaner than you found it. Remove unnecessary lines. Listen to the linter.
-->
Add CI Pipeline with Github Action, Deploy with sc-app-ci
## Links to Tickets or other pull requests

<!--
Base links to copy
- https://github.com/schul-cloud/schulcloud-server/pull/????
- https://ticketsystem.schul-cloud.org/browse/SC-????
-->
https://github.com/hpi-schul-cloud/schulcloud-server/pulls?q=is%3Apr+is%3Aopen+OPS-1559
https://ticketsystem.hpi-schul-cloud.org/browse/OPS-1874
## Changes

<!--
  What will the PR change?
  Why are the changes requiered?
  Short notice if a ticket exists, more detailed if not
-->
Enables CI to project with build and push to Dockerhub.
https://ticketsystem.hpi-schul-cloud.org/browse/OPS-1874
## Datasecurity <sub><sup>details [on Confluence](https://docs.schul-cloud.org/x/2S3GBg)</sup></sub>

<!--
  Notice about:
  - model changes
  - logging of user data
  - right changes
  - and other user data stuff
  If you are not sure if it is relevant, take a look at confluence or ask the data-security team.
-->
N/A
## Deployment

<!--
  Keep in mind to changes to seed data, if changes are done by migration scripts.
  Changes to the infrastructure have to discussed with the devops

  This point should includes following informations:
  - What is required for deployment?
  - Envirement variables like FEATURE_XY=true
  - Migration scripts to run, other requirements
-->
Deployment will be added in another task with sc-app-ci
https://ticketsystem.hpi-schul-cloud.org/browse/OPS-1559
## New Repos, NPM pakages or vendor scripts

<!--
  Keep in mind the stability, performance, activity and author.

  Describe why it is needed.
-->
N/A
## Screenshots of UI changes

<!--
  only needed for visual changes

  If visual changes exist, work together with UI/UX from beginning/ping UX with final PR
-->
N/A
## Approval for review

- All points were discussed with the ticket creator, support-team or product owner. The code upholds all quality guidelines from the PR-template.

> Notice: Please remove the WIP label if the PR is ready to review, otherwise nobody will review it.

### Link to Definiton of Done

More and detailed information on the _definition of done_ can be found [on Confluence](https://docs.schul-cloud.org/pages/viewpage.action?pageId=92831762)

mamutmk5 commented on .github/workflows/build_push_service.yml:
```diff
@@ -6,20 +6,169 @@ name: sc-app-build-push
+          text: ${{ env.BRANCH_NAME }}
+          regex: '^([a-zA-Z]+)\/[A-Z]+-[0-9]+-[a-zA-Z_]+$'
+          flags: gm
+      - name: Set env var JIRA BRANCH_PREFIX
+        run: |
+          echo "BRANCH_PREFIX is ${{ steps.branch_prefix.outputs.group1 }}"
+          echo BRANCH_PREFIX=${{ steps.branch_prefix.outputs.group1 }} >> $GITHUB_ENV
```
Why is it put to ENV vars if is at the output?

RalfNellesCap replied:
So other actions have access to env vars, we need this var to build correct dockertag with new naming convention.

mamutmk5 replied:
okay, then my question where not so good to understand.
Why do you put the output form an named step to the env vars again and not uesed the step output later?
The steps.branch_prefix.outputs.group1 is allrady an env var.....

mamutmk5 commented on .github/workflows/build_push_service.yml:
```diff
@@ -6,20 +6,169 @@ name: sc-app-build-push
+          text: ${{ env.BRANCH_NAME }}
+          regex: '^hotfix|feature\/([A-Z]+-[0-9]+)-[a-zA-Z_]+$'
+          flags: gm
+      - name: Set env var JIRA_TicketID
+        run: |
+          echo "JIRA_TicketID is ${{ steps.jira.outputs.group1 }}"
+          echo JIRA_TicketID=${{ steps.jira.outputs.group1 }} >> $GITHUB_ENV
```
Why is it put to ENV vars if is at the output?

RalfNellesCap replied:
Other actions have access to env vars, we need this var to build correct dockertag with new naming convention.

mamutmk5 commented on .github/workflows/build_push_service.yml:
```diff
@@ -6,20 +6,169 @@ name: sc-app-build-push
+        uses: juliangruber/read-file-action@v1
+        with:
+          path: ./version
+      - name: Set env AVCHECK_VERSION
+        run: |
+          echo "Version of antivirus_service is ${{ steps.read_version.outputs.content }}"
+          echo "AVCHECK_VERSION=${{ steps.read_version.outputs.content }}" >> $GITHUB_ENV
```
Why is it put to ENV vars if is at the output?

RalfNellesCap replied:
Other actions have access to env vars, we need this var to build correct dockertag with new naming convention.
see EPIC OPS-1499 Provide team based HPI Schul-Cloud instances
new naming convention confluence: Naming Convention on Builds https://docs.hpi-schul-cloud.org/display/TSC/Naming+Convention+on+Builds

mamutmk5 commented on .github/workflows/build_push_service.yml:
```diff
@@ -6,20 +6,169 @@ name: sc-app-build-push
+        run: echo Set BRANCH_NAME
+        env:
+          BRANCH_NAME: ${{ steps.branch.outputs.BRANCH_NAME }}
+      - name: Read version file
+        # The version to be used for tagging the images is read out of the version file
+        id: read_version
+        uses: juliangruber/read-file-action@v1
+        with:
+          path: ./version
```
Why is the version file read?

RalfNellesCap replied:
need for new naming convention dockertag master/ release

mamutmk5 replied:
what is with the version at setup.py and how do it use the ./version file?

https://github.com/hpi-schul-cloud/antivirus_check_service/blob/feature/OPS-1722-build_and_deploy_antivirus/setup.py

mamutmk5 commented on .github/workflows/build_push_service.yml:
```diff
@@ -6,20 +6,169 @@ name: sc-app-build-push
+          echo BRANCH_PREFIX=${{ steps.branch_prefix.outputs.group1 }} >> $GITHUB_ENV
+      # Same convention as implemented OPS-1664
+      # develop -> just develop_latest
+      - name: Assemble Dockertag for develop
+        if: env.BRANCH_PREFIX == 'develop'
+        run: echo DOCKER_TAG_POSTFIX=develop_latest >> $GITHUB_ENV
+      # feature -> feature_<JIRA ID>_latest
+      - name: Assemble Dockertag for feature
+        if: env.BRANCH_PREFIX == 'feature'
+        run: echo "DOCKER_TAG_POSTFIX=feature_${{env.JIRA_TicketID}}_latest" >> $GITHUB_ENV
+      # release -> release_v<AVCHECK_VERSION>_latest, release_v<AVCHECK_VERSION>_GITSHA
+      - name: Assemble Dockertag for release
+        if: env.BRANCH_PREFIX == 'release'
+        run: |
+          echo "DOCKER_TAG_POSTFIX=release_${{env.AVCHECK_VERSION}}_latest" >> $GITHUB_ENV
+          echo "DOCKER_TAG_SHA=release_${{env.AVCHECK_VERSION}}_${{env.GITHUB_SHA}}" >> $GITHUB_ENV
+      # hotfix -> hotfix_<JIRA ID>_latest
+      - name: Assemble Dockertag for hotfix
+        if: env.BRANCH_PREFIX == 'hotfix'
+        run: |
+          echo "DOCKER_TAG_POSTFIX=hotfix_${{env.JIRA_TicketID}}_latest" >> $GITHUB_ENV
+      # master -> master_v<AVCHECK_VERSION>_latest, master_v<AVCHECK_VERSION>_GITSHA
+      - name: Assemble Dockertag for master
+        if: env.BRANCH_PREFIX == 'master'
+        run: |
+          echo "DOCKER_TAG_POSTFIX=master_${{env.AVCHECK_VERSION}}_latest" >> $GITHUB_ENV
+          echo "DOCKER_TAG_SHA=master_${{env.AVCHECK_VERSION}}_${{env.GITHUB_SHA}}" >> $GITHUB_ENV
```
Too much special cases, why was they not designed as separate workflows?

RalfNellesCap replied:
See epic, to meet naming convention it is necessary.

RalfNellesCap replied:
See epic, to meet naming convention it is necessary. We will build all branches with one Workflow.
One thing was that actions have very little capability's on conditions. It doesn't look smart but now it works.

mamutmk5 commented on .github/workflows/build_push_service.yml:
```diff
@@ -6,20 +6,169 @@ name: sc-app-build-push
+          context: .
+          file: ./Dockerfile.scanfile
+          push: true
+          tags: ${{ env.DOCKER_NAMESPACE }}/${{ env.AVSCANFILE_REPO }}:${{ env.DOCKER_TAG_POSTFIX }}
```
why not more geneal why, with more tags then one pushed?

RalfNellesCap replied:
I see three steps for three docker images to be build. would be grateful for easier example.

mamutmk5 commented on .github/workflows/build_push_service.yml:
```diff
@@ -6,20 +6,169 @@ name: sc-app-build-push
+          file: ./Dockerfile.webserver
+          push: true
+          tags: ${{ env.DOCKER_NAMESPACE }}/${{ env.AVSCANWEBSERVER_REPO }}:${{ env.DOCKER_TAG_POSTFIX }}
+      - name: Build and push SHA Tag for Master and Release on Scanfile
+        if: ${{ github.event.inputs.branch-prefix == 'master' || github.event.inputs.branch-prefix == 'release' }}
+        uses: docker/build-push-action@v2.2.1
+        with:
+          context: .
+          file: ./Dockerfile.scanfile
+          push: true
+          tags: ${{ env.DOCKER_NAMESPACE }}/${{ env.AVSCANWEBSERVER_REPO }}:${{ env.DOCKER_TAG_SHA }}
+      - name: Build and push SHA Tag for Master and Release on Scanurl
+        if: ${{ github.event.inputs.branch-prefix == 'master' || github.event.inputs.branch-prefix == 'release' }}
+        uses: docker/build-push-action@v2.2.1
+        with:
+          context: .
+          file: ./Dockerfile.scanurl
+          push: true
+          tags: ${{ env.DOCKER_NAMESPACE }}/${{ env.AVSCANWEBSERVER_REPO }}:${{ env.DOCKER_TAG_SHA }}
+      - name: Build and push SHA Tag for Master and Release on Webserver
+        if: ${{ github.event.inputs.branch-prefix == 'master' || github.event.inputs.branch-prefix == 'release' }}
+        uses: docker/build-push-action@v2.2.1
+        with:
+          context: .
+          file: ./Dockerfile.webserver
+          push: true
+          tags: ${{ env.DOCKER_NAMESPACE }}/${{ env.AVSCANWEBSERVER_REPO }}:${{ env.DOCKER_TAG_SHA }}
```
To mutch steps with to offen build, why where the SHA tags not push with this bevore?

RalfNellesCap replied:
1. conditions in actions are limited. So 5 braches with 3 images.
2. we had SHA tags before an dev's and devops demanded to keep SHA Tags and enhanced epic / naming convention.

mamutmk5 commented on .github/workflows/build_push_service.yml:
```diff
@@ -6,20 +6,169 @@ name: sc-app-build-push
 on:
+  push:
+    branches:
+      - develop
+      - master
+      # release branches are of e.g. release/25.3.0
+      - release/[0-9]+.[0-9]+.[0-9]+
+      # OPS-1664 naming convention <branch>/<JIRA-Ticket ID>-<Jira_Summary>
+      - hotfix/[A-Z]+-[0-9]+-[a-zA-Z0-9]+_[a-zA-Z0-9]+*
+      - feature/[A-Z]+-[0-9]+-[a-zA-Z0-9]+_[a-zA-Z0-9]+*
```
Why are it on so many cases at once, it's created many special cases to handle.

RalfNellesCap replied:
No, Workflow should just fail on old naming scheme.

mamutmk5 commented on version:
```diff
@@ -0,0 +1 @@
+1.0.0
```
Why is a version file add what not used by python and why weren't the git tag used for versioning?

RalfNellesCap replied:
python should be the next step with our own python action for all sc-apps
See: feature/OPS-1722-build_and_deploy_antivirus

RalfNellesCap replied:
version file now is used by python the git tag uses the version text file for versioning.

mamutmk5 reviewed (changes requested):
On my opinion is to complex designed with too many special cases.

RalfNellesCap commented:
- Added SHA tag to all branches
- fixed routing through branches with new conditions to build dockertags
- fixed double versioning files, joint to single point of truth.

RalfNellesCap commented:
- unified version
- added SHA Tags to all branches
- cleaned up conditions-steps
- action runs
- pushes tag SHA
- pushes three images with 2 Tags each.

Asked for review and merge.
